### PR TITLE
[SPARK-14682][ML] Provide evaluateEachIteration method or equivalent for spark.ml GBTs

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

Provide evaluateEachIteration method or equivalent for spark.ml GBTs.

## How was this patch tested?

UT.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala [hunk 1]
```diff
@@ -365,6 +365,20 @@ class GBTClassifierSuite extends MLTest with DefaultReadWriteTest {
     assert(mostImportantFeature !== mostIF)
   }
 
+  test("model evaluateEachIteration") {
+    for (lossType <- Seq("logistic")) {
```
Comment (wangmiao1981) [inline mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: there is only one lossType. `for` is not necessary.

Comment (wangmiao1981) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: OK. It makes sense.

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala [hunk 2]
```diff
@@ -365,6 +365,20 @@ class GBTClassifierSuite extends MLTest with DefaultReadWriteTest {
...
+      val gbt = new GBTClassifier()
+        .setMaxDepth(2)
+        .setMaxIter(2)
+        .setLossType(lossType)
+      val model = gbt.fit(trainData.toDF)
+      val eval1 = model.evaluateEachIteration(validationData.toDF)
+      val eval2 = GradientBoostedTrees.evaluateEachIteration(validationData,
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: This is testing the spark.ml implementation against itself.  I was about to recommend using the old spark.mllib implementation as a reference.   However, the old implementation is not tested at all.  Would you be able to test against a standard implementation in R or scikit-learn (following the patterns used elsewhere in MLlib)?

File: mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala [hunk 3]
```diff
@@ -269,6 +269,21 @@ class GBTRegressionModel private[ml](
     new OldGBTModel(OldAlgo.Regression, _trees.map(_.toOld), _treeWeights)
   }
 
+  /**
+   * Method to compute error or loss for every iteration of gradient boosting.
+   *
+   * @param dataset Dataset for validation.
+   */
+  @Since("2.4.0")
+  def evaluateEachIteration(dataset: Dataset[_]): Array[Double] = {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala]: Do we want to support evaluation on other losses, as in the old API?  It might be nice to be able to without having to modify the Model's loss Param value.

**Review by jkbradley — COMMENTED** [review]
Thanks!  Just a few comments.

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala (hunk 2 above)
Comment (WeichenXu123) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: I search scikit-learn doc, there seems no similar method like `evaluateEachIteration`, we can only use `staged_predict` in `sklearn.ensemble.GradientBoostingRegressor` and then implement almost the whole logic again. In R package I also do not find this method.
Now I update the unit test, to compare with hardcoded result.

File: mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala [hunk 4]
```diff
@@ -269,6 +269,20 @@ class GBTRegressionModel private[ml](
     new OldGBTModel(OldAlgo.Regression, _trees.map(_.toOld), _treeWeights)
   }
 
+  /**
+   * Method to compute error or loss for every iteration of gradient boosting.
+   *
+   * @param dataset Dataset for validation.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala]: Add doc for "loss" arg, including what the options are

**Review by jkbradley — COMMENTED** [review]
For unit tests, what about this?
* Used a fixed random seed.
* Run for maxIter = 3
* Create models with 1 and 2 trees by manually getting the trees and constructing new GBT models.
* Check to make sure the loss for a model with 1 tree matches the first value returned by evaluateEachIteration for the other 2 models.
* Check to make sure the loss for a model with 2 trees matches the second value returned by evaluateEachIteration for the model with 3 trees.

**Review by jkbradley — COMMENTED** [review]
Just a tiny comment left.  Thanks!

**Comment by jkbradley** [comment]
LGTM
Merging with master
Thanks @WeichenXu123 !   Would you mind creating & linking a JIRA for the Python API update?
